File: frontend/lib/privacy.ts
import { apiFetch, type ApiFetchOptions } from "@/app/lib/http/client";
import { getBackendUrl } from "./env";
import type {
	AuditLogItem,
	BlockEntry,
	DeletionStatus,
	ExportStatus,
	NotificationPrefs,
	ProfilePrivacy,
} from "./types";

const BASE_URL = getBackendUrl();

async function request<T>(path: string, options: ApiFetchOptions = {}): Promise<T> {
	return apiFetch<T>(`${BASE_URL}${path}`, {
		cache: "no-store",
		...options,
	});
}

function authHeaders(userId: string, campusId: string | null): Record<string, string> {
	return {
		"X-User-Id": userId,
		...(campusId ? { "X-Campus-Id": campusId } : {}),
	};
}

export type PrivacyPatchPayload = Partial<Omit<ProfilePrivacy, "visibility">> & {
	visibility?: ProfilePrivacy["visibility"];
};

export type NotificationPatchPayload = Partial<NotificationPrefs>;

export type AuditLogPage = {
	items: AuditLogItem[];
	cursor?: number | null;
};

export async function fetchPrivacySettings(userId: string, campusId: string | null): Promise<ProfilePrivacy> {
	return request<ProfilePrivacy>("/settings/privacy", { headers: authHeaders(userId, campusId) });
}

export async function updatePrivacySettings(
	userId: string,
	campusId: string | null,
	patch: PrivacyPatchPayload,
): Promise<ProfilePrivacy> {
	return request<ProfilePrivacy>("/settings/privacy", {
		method: "PATCH",
		body: patch,
		headers: authHeaders(userId, campusId),
	});
}

export async function listBlocks(userId: string, campusId: string | null): Promise<BlockEntry[]> {
	return request<BlockEntry[]>("/privacy/blocks", { headers: authHeaders(userId, campusId) });
}

export async function blockUser(userId: string, campusId: string | null, targetId: string): Promise<BlockEntry> {
	return request<BlockEntry>(`/privacy/block/${targetId}`, {
		method: "POST",
		headers: authHeaders(userId, campusId),
	});
}

export async function unblockUser(userId: string, campusId: string | null, targetId: string): Promise<void> {
	await request<void>(`/privacy/block/${targetId}`, {
		method: "DELETE",
		headers: authHeaders(userId, campusId),
	});
}

export async function fetchNotificationPrefs(
	userId: string,
	campusId: string | null,
): Promise<NotificationPrefs> {
	return request<NotificationPrefs>("/settings/notifications", { headers: authHeaders(userId, campusId) });
}

export async function updateNotificationPrefs(
	userId: string,
	campusId: string | null,
	patch: NotificationPatchPayload,
): Promise<NotificationPrefs> {
	return request<NotificationPrefs>("/settings/notifications", {
		method: "PATCH",
		body: patch,
		headers: authHeaders(userId, campusId),
	});
}

export async function requestExportJob(userId: string, campusId: string | null): Promise<ExportStatus> {
	return request<ExportStatus>("/account/export/request", {
		method: "POST",
		headers: authHeaders(userId, campusId),
	});
}

export async function fetchExportStatus(userId: string, campusId: string | null): Promise<ExportStatus> {
	return request<ExportStatus>("/account/export/status", { headers: authHeaders(userId, campusId) });
}

export async function fetchExportDownload(userId: string, campusId: string | null): Promise<ExportStatus> {
	return request<ExportStatus>("/account/export/download", { headers: authHeaders(userId, campusId) });
}

export async function requestDeletion(userId: string, campusId: string | null): Promise<DeletionStatus> {
	return request<DeletionStatus>("/account/delete/request", {
		method: "POST",
		headers: authHeaders(userId, campusId),
	});
}

export async function confirmDeletion(
	userId: string,
	campusId: string | null,
	token: string,
): Promise<DeletionStatus> {
	return request<DeletionStatus>("/account/delete/confirm", {
		method: "POST",
		body: { token },
		headers: authHeaders(userId, campusId),
	});
}

export async function fetchDeletionStatus(userId: string, campusId: string | null): Promise<DeletionStatus> {
	return request<DeletionStatus>("/account/delete/status", {
		headers: authHeaders(userId, campusId),
	});
}

export async function fetchAuditLog(
	userId: string,
	campusId: string | null,
	params: { limit?: number; cursor?: number | null } = {},
): Promise<AuditLogPage> {
	const query = new URLSearchParams();
	if (params.limit) {
		query.set("limit", String(params.limit));
	}
	if (params.cursor) {
		query.set("cursor", String(params.cursor));
	}
	const suffix = query.toString() ? `?${query.toString()}` : "";
	return request<AuditLogPage>(`/account/audit${suffix}`, {
		headers: authHeaders(userId, campusId),
	});
}

File: backend/app/api/privacy.py
"""Privacy, notification, export, deletion, and audit endpoints."""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.domain.identity import (
	audit,
	deletion,
	export,
	notifications,
	policy,
	privacy,
	schemas,
)
from app.infra.auth import AuthenticatedUser, get_current_user

router = APIRouter()


def _map_policy_error(exc: policy.IdentityPolicyError) -> HTTPException:
	reason = exc.reason
	if isinstance(exc, policy.IdentityRateLimitExceeded):
		return HTTPException(status.HTTP_429_TOO_MANY_REQUESTS, detail=reason)
	if reason in {"block_exists"}:
		return HTTPException(status.HTTP_409_CONFLICT, detail=reason)
	if reason in {"block_missing", "user_missing", "export_not_found", "delete_not_requested"}:
		return HTTPException(status.HTTP_404_NOT_FOUND, detail=reason)
	if reason in {"block_self"}:
		return HTTPException(status.HTTP_400_BAD_REQUEST, detail=reason)
	if reason in {"delete_token_invalid"}:
		return HTTPException(status.HTTP_409_CONFLICT, detail=reason)
	return HTTPException(status.HTTP_400_BAD_REQUEST, detail=reason)


@router.get("/settings/privacy", response_model=schemas.PrivacySettings)
async def get_privacy(auth_user: AuthenticatedUser = Depends(get_current_user)) -> schemas.PrivacySettings:
	try:
		return await privacy.get_privacy_settings(auth_user.id)
	except policy.IdentityPolicyError as exc:
		raise _map_policy_error(exc) from None


@router.patch("/settings/privacy", response_model=schemas.PrivacySettings)
async def patch_privacy(
	payload: schemas.PrivacySettingsPatch,
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> schemas.PrivacySettings:
	try:
		return await privacy.update_privacy_settings(auth_user, payload)
	except policy.IdentityPolicyError as exc:
		raise _map_policy_error(exc) from None


@router.get("/privacy/blocks", response_model=list[schemas.BlockListEntry])
async def get_blocks(auth_user: AuthenticatedUser = Depends(get_current_user)) -> list[schemas.BlockListEntry]:
	return await privacy.list_blocks(auth_user.id)


@router.post("/privacy/block/{target_id}", response_model=schemas.BlockListEntry)
async def post_block(
	target_id: str,
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> schemas.BlockListEntry:
	try:
		return await privacy.block_user(auth_user, target_id)
	except policy.IdentityPolicyError as exc:
		raise _map_policy_error(exc) from None


@router.delete("/privacy/block/{target_id}")
async def delete_block(
	target_id: str,
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> dict[str, bool]:
	try:
		await privacy.unblock_user(auth_user, target_id)
	except policy.IdentityPolicyError as exc:
		raise _map_policy_error(exc) from None
	return {"unblocked": True}


@router.get("/settings/notifications", response_model=schemas.NotificationPreferences)
async def get_notifications(
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> schemas.NotificationPreferences:
	return await notifications.get_preferences(auth_user.id)


@router.patch("/settings/notifications", response_model=schemas.NotificationPreferences)
async def patch_notifications(
	payload: schemas.NotificationPreferencesPatch,
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> schemas.NotificationPreferences:
	return await notifications.update_preferences(auth_user.id, payload)


@router.post("/account/export/request", response_model=schemas.ExportStatus)
async def request_export(
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> schemas.ExportStatus:
	try:
		return await export.request_export(auth_user)
	except policy.IdentityPolicyError as exc:
		raise _map_policy_error(exc) from None


@router.get("/account/export/status", response_model=schemas.ExportStatus)
async def get_export_status(
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> schemas.ExportStatus:
	status_info = await export.get_status(auth_user.id)
	if not status_info:
		raise HTTPException(status.HTTP_404_NOT_FOUND, detail="export_not_found")
	return status_info


@router.get("/account/export/download", response_model=schemas.ExportStatus)
async def get_export_download(
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> schemas.ExportStatus:
	status_info = await export.get_status(auth_user.id)
	if not status_info:
		raise HTTPException(status.HTTP_404_NOT_FOUND, detail="export_not_found")
	if status_info.status != "ready":
		raise HTTPException(status.HTTP_409_CONFLICT, detail="export_pending")
	return status_info


@router.post("/account/delete/request", response_model=schemas.DeletionStatus)
async def post_delete_request(
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> schemas.DeletionStatus:
	try:
		return await deletion.request_deletion(auth_user)
	except policy.IdentityPolicyError as exc:
		raise _map_policy_error(exc) from None


@router.post("/account/delete/confirm", response_model=schemas.DeletionStatus)
async def post_delete_confirm(
	payload: schemas.DeletionConfirm,
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> schemas.DeletionStatus:
	try:
		return await deletion.confirm_deletion(auth_user, payload.token)
	except policy.IdentityPolicyError as exc:
		raise _map_policy_error(exc) from None


@router.post("/account/delete/force", response_model=schemas.DeletionStatus)
async def post_delete_force(auth_user: AuthenticatedUser = Depends(get_current_user)) -> schemas.DeletionStatus:
	"""Immediate, tokenless deletion for the authenticated user (development convenience)."""
	try:
		return await deletion.force_delete(auth_user)
	except policy.IdentityPolicyError as exc:
		raise _map_policy_error(exc) from None


@router.get("/account/delete/status", response_model=schemas.DeletionStatus)
async def get_delete_status(
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> schemas.DeletionStatus:
	try:
		return await deletion.get_status(auth_user.id)
	except policy.IdentityPolicyError as exc:
		raise _map_policy_error(exc) from None


@router.get("/account/audit", response_model=schemas.AuditLogPage)
async def get_audit_log(
	limit: int = Query(default=50, ge=1, le=policy.AUDIT_PAGE_MAX),
	cursor: int | None = Query(default=None, ge=1),
	auth_user: AuthenticatedUser = Depends(get_current_user),
) -> schemas.AuditLogPage:
	entries, next_cursor = await audit.fetch_audit_log(auth_user.id, limit=limit, cursor=cursor)
	items = [
		schemas.AuditLogItem(
			id=entry.id,
			event=entry.event,
			meta={k: str(v) for k, v in entry.meta.items()},
			created_at=entry.created_at,
		)
		for entry in entries
	]
	return schemas.AuditLogPage(items=items, cursor=next_cursor)
